### PR TITLE
Add support for SOCKS proxy

### PR DESCRIPTION
**Pull request by cketti**
Rebased pull request #861 on top of current `master`.
Made some additional changes.

The feature is currently disabled for release builds. We don't want to ship with incomplete proxy support because this might have privacy implications for our users.

### CODE REVIEW

File: k9mail-library/src/main/java/com/fsck/k9/mail/ssl/DefaultTrustedSocketFactory.java [hunk 1]
```diff
@@ -160,7 +169,18 @@ public Socket createSocket(Socket socket, String host, int port, String clientCe
...
+            if (proxySettings.enabled) {
+                InetSocketAddress proxyAddress = new InetSocketAddress(proxySettings.host, proxySettings.port);
+                Proxy proxy = new Proxy(Type.SOCKS, proxyAddress);
+
+                Socket underlying = new Socket(proxy);
+                InetSocketAddress serverAddress = new InetSocketAddress(host, port);
+                underlying.connect(serverAddress, SOCKET_CONNECT_TIMEOUT);
+
+                trustedSocket = socketFactory.createSocket(underlying, proxySettings.host, proxySettings.port, true);
+            } else {
```
Comment (philipwhiuk) [inline k9mail-library/src/main/java/com/fsck/k9/mail/ssl/DefaultTrustedSocketFactory.java]: The question of self-signed certificates for proxy implementations comes to mind. Should we be using our own socket factory here.

File: k9mail/src/main/java/com/fsck/k9/K9.java [hunk 2]
```diff
@@ -255,6 +256,9 @@
     private static boolean sMessageViewCopyActionVisible = false;
     private static boolean sMessageViewSpamActionVisible = false;
 
+    private static boolean sUseSocksProxy = false;
+    private static String sSocksProxyHost = "";
+    private static int sSocksProxyPort = 0;
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/K9.java]: It's a pity that the design has led to this class filling up with lots of constants. 

Would not make sense to just have

```private static ProxySettings sSocksProxySettings``` here instead.

**Comment by cketti** [comment]
This branch currently only supports using a proxy for TLS connections. See #1192 for an attempt to fix that.

File: k9mail-library/src/main/java/com/fsck/k9/mail/ssl/DefaultTrustedSocketFactory.java [hunk 3]
```diff
@@ -160,7 +169,18 @@ public Socket createSocket(Socket socket, String host, int port, String clientCe
...
-            trustedSocket = socketFactory.createSocket();
+            if (proxySettings.enabled) {
+                InetSocketAddress proxyAddress = new InetSocketAddress(proxySettings.host, proxySettings.port);
+                Proxy proxy = new Proxy(Type.SOCKS, proxyAddress);
+
+                Socket underlying = new Socket(proxy);
+                InetSocketAddress serverAddress = new InetSocketAddress(host, port);
+                underlying.connect(serverAddress, SOCKET_CONNECT_TIMEOUT);
+
+                trustedSocket = socketFactory.createSocket(underlying, proxySettings.host, proxySettings.port, true);
```
Comment (philipwhiuk) [inline k9mail-library/src/main/java/com/fsck/k9/mail/ssl/DefaultTrustedSocketFactory.java]: This feels wrong:

You create a plain Socket with a proxy.
You connect that plain Socket to the end point.
Then you create (and connect, but this is hidden by the SSL API) that SSL socket wrapped round that socket to the proxy.

Don't you now have a connection to the proxy via the server via the proxy?

i.e

`Client -> Proxy -> Server -> Proxy`

Comment (philipwhiuk) [reply k9mail-library/src/main/java/com/fsck/k9/mail/ssl/DefaultTrustedSocketFactory.java]: I think we want:

```
                InetSocketAddress proxyAddress = new InetSocketAddress(proxySettings.host, proxySettings.port);
                Proxy proxy = new Proxy(Type.SOCKS, proxyAddress);
                Socket underlying = new Socket(proxy);
                InetSocketAddress serverAddress = new InetSocketAddress(host, port);
                underlying.connect(serverAddress, SOCKET_CONNECT_TIMEOUT);
                trustedSocket = socketFactory.createSocket(underlying, host, port, true);
```

That is, connect a socket to SOCKS PROXY.
Use that underlying socket to connect to the remote socket.
Upgrade the remote socket proxied connection to SSL.

**Pull request closed without merging**